runtime/server/metrics_views.go

```diff
@@ -78,89 +58,223 @@ func (s *Server) MetricsViewToplist(ctx context.Context, req *api.MetricsViewTop
+	// NOTE: Partially mocked - timestamp column is hardcoded
 
 	sql, args, err := buildMetricsTimeSeriesSQL(req)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "error building query: %s", err.Error())
 	}
 
-	rows, err := s.query(ctx, req.InstanceId, &drivers.Statement{
-		Query: sql,
-		Args:  args,
-	})
+	meta, data, err := s.runQuery(ctx, req.InstanceId, sql, args)
 	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+		return nil, err
 	}
-	defer rows.Close()
 
-	meta, err := rowsToMeta(rows)
+	resp := &api.MetricsViewTimeSeriesResponse{
+		Meta: meta,
+		Data: data,
+	}
+
+	return resp, nil
+}
+
+// MetricsViewTotals implements RuntimeService
+func (s *Server) MetricsViewTotals(ctx context.Context, req *api.MetricsViewTotalsRequest) (*api.MetricsViewTotalsResponse, error) {
+	sql, args, err := buildMetricsTotalsSql(req)
 	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+		return nil, status.Errorf(codes.Internal, "error building query: %s", err.Error())
 	}
 
-	data, err := rowsToData(rows)
+	meta, data, err := s.runQuery(ctx, req.InstanceId, sql, args)
 	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+		return nil, err
 	}
 
-	resp := &api.MetricsViewTimeSeriesResponse{
+	resp := &api.MetricsViewTotalsResponse{
 		Meta: meta,
-		Data: data,
+		Data: data[0],
 	}
 
 	return resp, nil
 }
 
-// MetricsViewTotals implements RuntimeService
-func (s *Server) MetricsViewTotals(ctx context.Context, req *api.MetricsViewTotalsRequest) (*api.MetricsViewTotalsResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method not implemented")
+func (s *Server) runQuery(ctx context.Context, instanceId string, sql string, args []any) ([]*api.SchemaColumn, []*structpb.Struct, error) {
+	rows, err := s.query(ctx, instanceId, &drivers.Statement{
+		Query: sql,
+		Args:  args,
+	})
+	if err != nil {
+		return nil, nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	defer rows.Close()
+
+	meta, err := rowsToMeta(rows)
+	if err != nil {
+		return nil, nil, status.Error(codes.Internal, err.Error())
+	}
+
+	data, err := rowsToData(rows)
+	if err != nil {
+		return nil, nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return meta, data, nil
 }
 
 func buildMetricsTimeSeriesSQL(req *api.MetricsViewTimeSeriesRequest) (string, []any, error) {
-	timeField := "time"
-	timeCol := fmt.Sprintf("DATE_TRUNC(%s, %s) AS %s", timeField, req.TimeGranularity, timeField)
+	// TODO: get from Catalog
+	timeField := "timestamp"
+	timeCol := fmt.Sprintf("DATE_TRUNC('%s', %s) AS %s", req.TimeGranularity, timeField, timeField)
 	selectCols := append([]string{timeCol}, req.MeasureNames...)
 
-	whereClause := "time >= ? AND time < ? "
-	args := []any{time.UnixMilli(req.TimeStart), time.UnixMilli(req.TimeEnd)}
+	whereClause := fmt.Sprintf("%s >= epoch_ms(?) AND %s < epoch_ms(?) ", timeField, timeField)
+	args := []any{req.TimeStart, req.TimeEnd}
```

Comment:
Is passing time struct a problem on the duckdb side? It seems like an API concern to parse request types correctly (later, we might add timestamp de-serialization to the protobufs – e.g. there's the [`google.protobuf.Timestamp`](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#timestamp) type)

Reply:
It was not passing time struct. It was just passing a number. Hence the `epoch_ms` wrapping.